src/main/java/org/paukov/combinatorics3/KPermutationGenerator.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Combinatorics Library 3
+ * Copyright 2009-2016 Dmytro Paukov d.paukov@gmail.com
+ */
+package org.paukov.combinatorics3;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * This generator generates k-permutations of the specified initial vector
+ * <p>
+ * A k-permutation is an ordering of a subset in the subset of all possible
+ * orderings. For example, the set containing 2 digits of 123, has
+ * six permutations: 12, 21, 13, 31, 23 and 32.
+ * <p>
+ * This is an example of the k-permutations of 3 string items (apple, orange,
+ * cherry):
+ * <p>
+ * <blockquote>
+ *
+ * <pre>
+ *
+ *     List<List<String>> permutations =
+ *             Generator.permutation(Arrays.asList("apple", "orange", "cherry"))
+ *                     .k(2)
+ *                     .stream()
+ *                     .collect(toList());
+ *     permutations.stream().forEach(System.out::println);
+ *
+ * </pre>
+ *
+ * </blockquote>
+ * <p>
+ * And the result
+ * <p>
+ * <blockquote>
+ *
+ * <pre>
+ *   [apple, orange]
+ *   [orange, apple]
+ *   [apple, cherry]
+ *   [cherry, apple]
+ *   [orange, cherry]
+ *   [cherry, orange]
+ * </pre>
+ *
+ * </blockquote>
+ * <p>
+ *
+ * @param <T> Type of the elements in the permutations
+ * @author Charlie Feng
+ * @version 3.4
+ */
+class KPermutationGenerator<T> implements IGenerator<List<T>> {
+
+  final List<T> originalVector;
+  final int length;
+
+  /**
+   * Constructor
+   *
+   * @param vector Vector which is used for k-permutation generation
+   * @param length number of elements in k-permutation
+   */
+  KPermutationGenerator(Collection<T> vector,
+                        int length) {
+    this.originalVector = new ArrayList<>(vector);
+    this.length = length;
+  }
+
+  @Override
+  public Iterator<List<T>> iterator() {
+    return  Generator.combination(originalVector)
+            .simple(length)
+            .stream()
+            .flatMap(combination -> Generator.permutation(combination).simple().stream() )
+            .distinct()
```

Comment:
Line #83

I wonder if we should call this .distinct() method here? I think the KPermutationGenerator should return the permutations regardless of the duplicates in the original combinations. Users of the KPermutationGenerator can call .distinct() explicitly in their code only if it is really needed. The library will not always perform unnecessary distinct() operation and let users decide what to do with the duplicates. What do you think?

```
List<List<String>> permutations =
            Generator.permutation("a", "a", "b")
                    .k(2)
                    .stream()
                    .distinct()                    // <- explicitly removes duplicates
                    .collect(toList());
```

If no duplicates present:

```
List<List<String>> permutations =
            Generator.permutation("a", "b", "c")
                    .k(2)
                    .stream()
                    .collect(toList());
```

Reply:
When the input vector do not contains duplicated element, k-permutation return same result, with or without "distinct()" line

When the input vector contains duplicated element, I give 4 examples that return different result.

The result of Example 1 looks weird, it return 5 elements.


Example 1
Generator.permutation("a", "a", "b").k(2)
	With distinct, return
	["a", "a"],["a", "b"],["b", "a"]
	Without distinct, return
	['a', 'a'], ['a', 'b'], ['b', 'a'], ['a', 'b'], ['b', 'a']

Example 2
Generator.permutation("a", "a", "b").k(1)
	With distinct, return
	['a'], ['b']
	Without distinct, return
	['a'], ['a'], ['b']

Example 3
Generator.permutation("a", "a", "a").k(2)
	With distinct, return
	['a', 'a']
	Without distinct, return
	['a', 'a'], ['a', 'a'], ['a', 'a']

Example 4
Generator.permutation(Arrays.asList(1, 2, 2, 3)).k(3)
	With distinct(), return 
	[1, 2, 2],[2, 1, 2],[2, 2, 1],[1, 2, 3],[1, 3, 2],[3, 1, 2],[3, 2, 1],[2, 3, 1],[2, 1, 3],[2, 2, 3],[2, 3, 2],[3, 2, 2],
	Without distinct(), return
	[1, 2, 2], [2, 1, 2], [2, 2, 1], [1, 2, 3], [1, 3, 2], [3, 1, 2], [3, 2, 1], [2, 3, 1], [2, 1, 3], [1, 2, 3], [1, 3, 2], [3, 1, 2], [3, 2, 1], [2, 3, 1], [2, 1, 3], [2, 2, 3], [2, 3, 2], [3, 2, 2]


Then I tried changed from 
            .flatMap(combination -> Generator.permutation(combination).simple().stream())
            .distinct()
into 
            .flatMap(combination -> Generator.permutation(combination).simple(PermutationGenerator.TreatDuplicatesAs.IDENTICAL).stream())

I get ['a', 'a'], ['a', 'a'], ['a', 'a'],['a', 'a'], ['a', 'a'], ['a', 'a'] in example 3


So which style do you recommended?
Style 1: 
            .flatMap(combination -> Generator.permutation(combination).simple().stream())
            .distinct()
Style 2:
            .flatMap(combination -> Generator.permutation(combination).simple().stream())
Style 3:
            .flatMap(combination -> Generator.permutation(combination).simple(PermutationGenerator.TreatDuplicatesAs.IDENTICAL).stream())

Reply:
I see two options here:

1) Removing the `distinct()` from the generator (line #83). It will produce "normal" permutations if an input vector does not have duplicates (this is expected) and it will generate the "weird" permutations for input vectors that contain duplicates like the examples above (1-4). I think these "weird" results are expected too. This is because the input vectors don't contain unique values and the combinations are not unique as the result of that. Moreover I think the generator of the k-permutations should always return expected number of elements based on the formula: https://en.wikipedia.org/wiki/Permutation#k-permutations_of_n. The formula does not include any duplicates. So I think having duplicates in the input vector should be considered as a special case and treated accordingly by calling the `distinct()` only if it is really needed. For example, the first example will be "fixed" by adding `.distinct()` before `.collect()` somewhere in a client's code (not in the generator's code). Other examples can be "fixed" similarly. But user should explicitly do that rather than we silently include that `.distinct()` into our generator and will always do it even if it is not needed (no duplicates in the input vector) potentially reducing the performance.

If we choose this option we will not need this test as it doesn't test the generator, but tests the `.distinct()` call.

```
@Test
public void test_identical_k2_permutation() {
  List<List<String>> permutations =
          Generator.permutation("a", "a", "b")
                  .k(2)
                  .stream()
                  .distinct()              // <- explicitly removes duplicates
                  .collect(toList());

  assertThat(permutations).hasSize(3);
  System.out.println("Identical k2-permutations of (a, a, b):");
  permutations.stream().forEach(System.out::println);
  assertThat(permutations.get(0)).containsExactly("a", "a");
  assertThat(permutations.get(1)).containsExactly("a", "b");
  assertThat(permutations.get(2)).containsExactly("b", "a");
}
```

2) Keeping the `distinct()` as it is there right now. In this case we will perform unnecessary actions for the vectors of the unique elements (I expect these are the majority of the cases). If we choose that option I think we should highlight it somehow, so users are not surprised when our generator in the example 4 returns 12 permutations instead of the calculated 18 according to the formula https://en.wikipedia.org/wiki/Permutation#k-permutations_of_n. We have 12 because of the duplicates. We can add word 'Distinct' somewhere to the class name and the corresponding methods javadocs (for example, `KPermutationDistinctGenerator`).  Also as the real source of the duplicates is the combinations, we should call  `.distinct()` before `.flatMap()`:

```
  @Override
  public Iterator<List<T>> iterator() {
    return  Generator.combination(originalVector)
            .simple(length)
            .stream()
            .distinct()
            .flatMap(combination -> Generator.permutation(combination).simple().stream() )
            .iterator();
  }
```

Personally I prefer the first option of removing the `distinct()` from the generator, but I am ok if you think we should keep it there. In that case we should follow case 2 and rename the class to something like KPermutationDistinctGenerator. What do you prefer?

Reply:
Sorry I've forgotten that the `Generator.permutation(combination).simple()` can automatically use the `DuplicatedPermutationIterator` if a combination has duplicates. So I think we should follow that pattern and implement the option 2 as you originally suggested, but put `.distinct()` before `.flatMap()`. Users can still use a combination of both the Combination and Permutation generators if they need to ignore the duplicates https://github.com/dpaukov/combinatoricslib3#5-k-permutations

Reply:
How about we provide an optional parameter into k() method:

* without optional parameter:
Generator.permutation("a", "b", "c")
                  .k(2)
The implementation will check whether input parameter contains duplicated element. if duplication found, throw exception something like "Found duplicated element xxx in input element"
And the implemantion will not call "distinct()", because the useless distinct() call may consume lots of CPU and memory if the input vector is large.


* with optional parameter:
Generator.permutation("a", "a", "c")
                  .k(2, ALLOW_IDENTICAL_INPUT) //you may suggest the name of enum, I'm not good at English.
The implementation will ALLOW  input parameter contains duplicated element. and distinct() will be called to remove duplicated result.

* with optional parameter:
Generator.permutation("a", "a", "c")
                  .k(2, ALLOW_IDENTICAL_OUTPUT) //you may suggest the name of enum, I'm not good at English.
The implementation will ALLOW  input parameter contains duplicated element. and distinct() will NOT be called, duplicated result could be exist. 
duplicated result could make sense under some context. for example

Generator.permutation("Boy", "Boy", "Boy", "Girl", "Girl")
                  .k(2, ALLOW_IDENTICAL_OUTPUT)
                  .filer( MUST_HAVE_BOY)
                  .count()
will answer the question: "3 boys and 2 girls pick up a team leader and a vice team leader, at least one boy should be pick, how many possible permutation?"

Reply:
Code updated, please review changed code.